File: src/DFS/ConstructBinaryTreeInorderPostorderTraversal.java
package DFS;

import java.util.HashMap;
import java.util.Map;

// https://leetcode.com/problems/construct-binary-tree-from-inorder-and-postorder-traversal/
public class ConstructBinaryTreeInorderPostorderTraversal {
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    int index = 0;
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        if (inorder.length != postorder.length || inorder.length == 0 || postorder.length == 0) {
            return null;
        }
        Map<Integer, Integer> inOrderMap = new HashMap<>();
        for (int i = 0; i < inorder.length; i++) {
            inOrderMap.put(inorder[i], i);
        }
        index = postorder.length - 1;
        return build(postorder, inOrderMap, 0, index);
    }


    private TreeNode build(int[] postOrder, Map<Integer, Integer> inOrderMap, int left, int right) {
        if (left > right) {
            return null;
        }
        int rootValue = postOrder[index--];
        TreeNode newRoot = new TreeNode(rootValue);
        int cut = inOrderMap.get(rootValue);
        newRoot.right = build(postOrder, inOrderMap, cut + 1, right);
        newRoot.left = build(postOrder, inOrderMap, left, cut - 1);
        return newRoot;
    }

    public void initialize() {
        int[] postorder = {9,15,7,20,3};
        int[] inorder = {9,3,15,20,7};

        preOrder(buildTree(inorder, postorder));
    }

    public void preOrder(TreeNode node) {
        if (node == null) {
            return;
        }
        System.out.print(node.val + ", ");
        preOrder(node.left);
        preOrder(node.right);
    }

    public static void main(String[] args) {
        ConstructBinaryTreeInorderPostorderTraversal mdt = new ConstructBinaryTreeInorderPostorderTraversal();
        mdt.initialize();
    }
}

File: py/ConstructBinaryTreeInorderPostorderTraversal.py
# https://leetcode.com/problems/construct-binary-tree-from-inorder-and-postorder-traversal/description/

from collections import defaultdict
from typing import List, Optional
from tree_node import *


class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        valueToIndex, self.postIndex = defaultdict(int), len(postorder) - 1
        for index, value in enumerate(inorder):
            valueToIndex[value] = index

        def dfs(left: int, right: int):
            if left > right:
                return None
            rootValue = postorder[self.postIndex]
            self.postIndex -= 1
            cut = valueToIndex[rootValue]
            rootNode = TreeNode(rootValue)
            rootNode.right = dfs(
                cut + 1, right
            )  # must do right first because postorder iterate right then left when postIndex move from len - 1 to 0
            rootNode.left = dfs(left, cut - 1)

            return rootNode

        return dfs(0, self.postIndex)


s = Solution()
inorder = [9, 3, 20]
postorder = [9, 20, 3]
preOrder(s.buildTree(inorder, postorder))
